File: maximum path sum.cpp
# time: O(n) and space: O(1)
class Solution {
public:
    int solve(TreeNode* root,int &result) {
        if (!root){
            return 0;
        }
        // maximum value will not contibute in max path sum
        // adding negative value to another value[positive or negative] will always decrease value
        int left = max(0, solve(root->left, result));
        int right = max(0, solve(root->right, result));
        // consireding curve path
        result = max(result, root->val + left + right);
        // considering left or right path
        return root->val + max(left, right);
    }
    int maxPathSum(TreeNode* root){
        int result = INT_MIN;
        solve(root, result);
        return result;
    }
};

File: maximum path sum.py
class Solution:
    def __init__(self):
        self.result = float('-inf')
    def solve(self, root):
        if root is None:
            return 0
        # negative value will naver help to  to increase sum
        left = max(0, self.solve(root.left))
        right = max(0, self.solve(root.right))
        # taking curve path
        self.result = max(self.result, root.val + left + right)
        # taking left or right path
        return root.val + max(left, right)
        
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.solve(root)
        return self.result
        
